### DISCUSSION TITLE
Unwanted Updates when Installing From Updatesite

### DISCUSSION THREAD

**Discussion by tgoethel** [eclipse-platform/eclipse.platform.ui#1127] [Q&A]
Dear Eclipse community,
I am working with Eclipse 2023-06 (Version: 2023-06 (4.28.0), Build id: 20230608-1333), currently. Recently, I installed an RCP feature, say MyFeature, using an additional updatesite and experienced unexpected behaviour. At the time of installation of this feature, I had version 5.0.1 of google.inject installed in Eclipse.

![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/cc690201-ea0b-4ffb-abaa-e70d75755ea4)

To reproduce the behaviour, I installed 'MyFeature' in two ways in separate Eclipse instances:
1. When I install it with checkbox "Contact all updatesites ..." disabled, installation is successful and google.inject is still at version 5.0.1
 
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/19054ca9-4f8d-4e0c-bfcd-9114a19e9e92)
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/38a9fa9d-ad22-447f-b777-400acc4203ca)

2. When I install it with checkbox "Contact all updatesites ..." enabled, installation is successful, but also google.inject is updated to version 7.0.0.
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/7269e7df-dd79-458d-9f58-e5018d07851a)
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/42312693-0fb5-4bfa-a150-27b16f80880f)

I checked that MyFeature is really only available on one updatesite.

As google.inject is not needed to be upgraded as Way 1 shows, I would expect Way 2 to also leave google.inject as it is.

Is my expectation wrong or is there a bug in Eclipse?

Thanks and best wishes,
Thomas

**Comment by merks** [comment]
We don't know what you have installed **and** we don't know which update sites are available for updates.  Probably this one is available:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/7adea83f-45bc-40c2-bf34-686d62cde354)

That site does have com.google.inject version 7.0.0.

There are a bunch of things you may have installed and that may need updating that require that higher version:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/8c00667a-aa1e-4b52-ae27-02bd0d94284b)

**Reply by tgoethel** [reply]
When I install "MyFeature" with the checkbox disabled, the installation is successful, because all dependencies are satisifed. Only when the checkbox is enabled, the unnecessary update of google.inject is performed.

I can try to provide more information if necessary. My question is more of a general nature: Is it expected behaviour that Eclipse performs unnecessary implicit updates?

**Reply by ingomohr** [reply]
Thank you for the update, @merks! Can you second that if the checkbox "Contact all updatesites" is checked plugins in the dependency tree are being updated - vs not updated if that checkbox is not checked?

**Reply by tgoethel** [reply]
(I would actually not have any issues with implicit unnecessary updates. However, google.inject 7.0.0 has a dependency to sun.misc, which I do not have installed)

**Reply by merks** [reply]
It's not clear from the details provided if the update is unnecessary.  That dependency comes form the JRE:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/4f2b29ed-56d8-4bf2-866d-21bec02cee37)

In any case, p2 does have a tendency to install the latest version of things, depending of course on the version range constraints on that dependency.

**Reply by tgoethel** [reply]
If it was a necessary update, wouldn't it be necessary as well when disabling the checkbox "Contact all updatesites ...", thus, resulting in non-successful installation of "MyFeature"?

**Reply by merks** [reply]
In general, p2 will take into account the requirements of your feature (the thing(s) being installed or updated) along with the entire set of installable units IUs available from all enabled update sites and will then create a solution that satisfies all those requirements.  And, as I said, p2 has a tendency to prefer newer versions over older versions.  So yes, it might well update bundles that do not strictly require an update.

**Reply by tgoethel** [reply]
Thanks for the clarification!

**Comment by laeubi** [comment]
P2 uses a [SAT solver](https://en.wikipedia.org/wiki/SAT_solver) and as SAT is a NP-hard problem usually algorithms implementing such only guarantee that they find **ONE** solution but not necessarily **THE** solution in regard to some not formulated constraints, so for a given input there might be _different_ options that satisfy the problem.

So if you check `Contact all updatesites` you effectively adding more things a solver can choose and that might give different results (even though for that kind of problem if they use a randomized start it is even not guaranteed that the same _input_ always give the same _output_).

So given what you provided, it seems at least *valid* to use version 7 here over version 5, so the question is more why is this update "unwanted"? Just because one *could* find another solution? If yes, this is not a bug (in eclipse).
If it is unwanted because something now *fails* this is not a bug in Eclipse but in the provided metadata of the failing code (e.g. it should have contained its package imports to not include the next major version).

**Reply by tgoethel** [reply]
Thanks for the clarification! 

The update of google.inject to version 7.0.0 is not explicitly prohibited by any upper version constraint. The update is unexpected and unwanted, nevertheless, because only google.inject 7.0.0 has a dependency to sun.misc. This plugin is not installed in my Eclipse. So when I use a run configuration with google.inject, it does not work anymore, because sun.misc is missing.

**Reply by tgoethel** [reply]
My experience is more or less as follows:
- I install MyFeature, because I want to use it in my run configuration.
- As I am lazy, I leave the checkbox "Contact all updatesites ..." enabled like I've done throughout the last couple of years.
- Eclipse implicitly updates google.inject, which is incompatible with my Eclipse setup.
- Running my run configuration does not work.

When I replace my laziness with unchecking the checkbox "Contact all updatesites ...", everything works just fine.

**Reply by laeubi** [reply]
> So when I use a run configuration with google.inject, it does not work anymore, because sun.misc is missing

Does it really not working or does it gives a validation error but you can still proceed?

**Reply by tgoethel** [reply]
It gives me a validation error:
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/145323738/09652781-39c2-4f75-ab6e-c522018e5af8)

**Reply by laeubi** [reply]
That should be fixed in the latest eclipse release (2023-09) and you can just ignore the problem and proceed in older versions.

**Reply by tgoethel** [reply]
Ok, thank you very much!

**Reply by laeubi** [reply]
> I get an error, because some features cannot be loaded that use google.inject.

Can you *provide* the error maybe? Its really hard to guess whats going wrong otherwise. Also make sure to at best use the latest Eclipse Release (2023-09) when possible.

**Reply by tgoethel** [reply]
I am very sorry, but I was too quick when posting my previous post that I also deleted in the meantime. When ignoring the validation error, things work nevertheless as expected.

**Reply by merks** [reply]
That's what I thought because sun.misc is present in a normal JDK/JRE...
